src/extend.cpp

```diff
@@ -127,48 +133,143 @@ sparse_coo extend(const sparse_coo &sparse_rr_mat,
-      float dist = rr_dists[*rr_it];
-      if (j == i) {
-        continue;
+      long unique_neighbors = 0;
+      float prev_value = -1;
+      bool new_val;
+      for (auto rr_it = rr_ordered_idx.cbegin(); rr_it != rr_ordered_idx.cend();
+           ++rr_it) {
+        long j = std::get<1>(sparse_rr_mat)[row_start_idx[i] + *rr_it];
+        float dist = rr_dists[*rr_it];
+        if (j == i) {
+          continue;
+        }
+        if (unique_neighbors < kNN) {
+          dists[i].push_back(dist);
+          i_vec[i].push_back(i);
+          j_vec[i].push_back(j);
+          if (count_unique_distances)
+            new_val = abs(dist - prev_value) >= epsilon;
+            if (new_val)
+            {
+              unique_neighbors++;
+              prev_value = dist;
+            }
+          else
+          {
+            unique_neighbors = j_vec[i].size();
+          }
+        } else {
+          break; // next i
+        }
       }
-      bool new_val = abs(dist - prev_value) >= epsilon;
-      if (unique_neighbors < kNN || !new_val) {
-        dists.push_back(dist);
-        i_vec.push_back(i);
-        j_vec.push_back(j);
-        if (new_val) {
-          unique_neighbors++;
-          prev_value = dist;
+    }
+    len += dists[i].size();
+  }
+  // Combine the lists from each thread
+  std::vector<float> dists_all = combine_vectors(dists, len);
+  std::vector<long> i_vec_all = combine_vectors(i_vec, len);
+  std::vector<long> j_vec_all = combine_vectors(j_vec, len);
+  // Only count reciprocal matches
+  if (reciprocal_only)
+  {
+    std::vector<std::vector<float>> filtered_dists(n_samples);
+    std::vector<std::vector<long>> filtered_i_vec(n_samples);
+    std::vector<std::vector<long>> filtered_j_vec(n_samples);
+    len = 0;
+#pragma omp parallel for schedule(static) num_threads(num_threads) reduction(+:len)
+    for (long x = 0; x < i_vec_all.size(); x++)
+    {
+      if (i_vec_all[x] < j_vec_all[x])
+      {
+          for (long y = 0; y < i_vec_all.size(); y++)
+          {
+              if (i_vec_all[x] == j_vec_all[y] && j_vec_all[x] == i_vec_all[y])
+              {
+                  filtered_dists[i_vec_all[x]].push_back(dists_all[x]);
+                  filtered_i_vec[i_vec_all[x]].push_back(i_vec_all[x]);
+                  filtered_j_vec[i_vec_all[x]].push_back(j_vec_all[x]);
+                  break;
+              }
+          }
+      }
+      len += filtered_dists[i_vec_all[x]].size();
+    }
+    // Combine the lists from each thread
+    std::vector<float> filtered_dists_all = combine_vectors(filtered_dists, len);
+    std::vector<long> filtered_i_vec_all = combine_vectors(filtered_i_vec, len);
+    std::vector<long> filtered_j_vec_all = combine_vectors(filtered_j_vec, len);
+    return (std::make_tuple(filtered_i_vec_all, filtered_j_vec_all, filtered_dists_all));
+  }
+  else
+  {
+    return (std::make_tuple(i_vec_all, j_vec_all, dists_all));
```

Comment:
I don't like two returns in longer functions as it makes control harder to follow -- you could instead change lines 226-228 to overwrite i_vec_all etc

Reply:
Done in ac73fb7